### ISSUE TITLE
[Performance] Enhanced CPU Quantization: Add Offset Support for Improved Accuracy

### ISSUE THREAD

**Issue by EffortlessSteven** [EffortlessMetrics/BitNet-rs#418]
## Problem Description

The current CPU quantization functions `quantize_value` and `dequantize_value` in `crates/bitnet-quantization/src/utils.rs` use simplified linear quantization that only supports scale factors. This limits quantization accuracy for tensors with non-zero means, as optimal quantization often requires both scale and offset (zero-point) parameters for asymmetric quantization schemes.

## Environment

- **Component**: `bitnet-quantization` crate, CPU quantizer utilities
- **Files**: `crates/bitnet-quantization/src/utils.rs` (lines 95-119)
- **Build Configuration**: `--no-default-features --features cpu`
- **Related Features**: I2_S, TL1, TL2 quantization methods

## Current Implementation Analysis

**Current Functions:**
```rust
pub fn quantize_value(value: f32, scale: f32, bits: u8) -> i8 {
    let max_val = (1 << (bits - 1)) - 1;
    let min_val = -(1 << (bits - 1));

    if value.is_finite() && scale != 0.0 && scale.is_finite() {
        let normalized = value / scale;  // Simple linear scaling
        let quantized = normalized.round() as i32;
        return quantized.clamp(min_val, max_val) as i8;
    }
    0i8
}

pub fn dequantize_value(quantized: i8, scale: f32) -> f32 {
    if scale.is_finite() {
        quantized as f32 * scale  // Simple linear scaling
    } else {
        0.0
    }
}
```

**Limitations:**
1. **No Offset Support**: Current implementation assumes zero-point is always 0, which is suboptimal for asymmetric distributions
2. **Reduced Accuracy**: Without offset, quantization error increases for tensors with non-zero mean values
3. **Limited Compatibility**: Many modern quantization schemes (e.g., GGML's asymmetric quantization) require offset parameters
4. **Performance Impact**: Suboptimal quantization affects model accuracy, especially for activation tensors

## Root Cause Analysis

The current implementation uses symmetric quantization (`Q = round(value / scale)`) which assumes the data distribution is centered around zero. For real neural network tensors with non-zero means, asymmetric quantization (`Q = round(value / scale) + offset`) provides:

1. **Better Range Utilization**: Full quantization range usage for skewed distributions
2. **Reduced Quantization Error**: Lower MSE for non-centered data
3. **Industry Compatibility**: Alignment with GGML, ONNX, and TensorFlow quantization standards

## Impact Assessment

**Severity**: Medium-High
- **Affected Components**: All CPU quantization operations (I2_S, TL1, TL2)
- **Performance Impact**: Suboptimal model accuracy due to increased quantization error
- **Compatibility Impact**: Limited interoperability with industry-standard quantization formats
- **User Impact**: Reduced inference quality for models requiring asymmetric quantization

**Usage Analysis:**
- Used by TL2 quantizer for dequantization (line 405 in `tl2.rs`)
- Referenced in comprehensive mutation tests (`scale_factor_boundary_mutation_killers.rs`)
- Core utility for all CPU-based quantization operations

## Proposed Solution

Enhance the quantization utilities to support optional offset parameters while maintaining backward compatibility:

### Enhanced Function Signatures

```rust
/// Enhanced quantization with optional offset support
#[inline]
pub fn quantize_value_with_offset(value: f32, scale: f32, offset: f32, bits: u8) -> i8 {
    let max_val = (1 << (bits - 1)) - 1;
    let min_val = -(1 << (bits - 1));

    if value.is_finite() && scale != 0.0 && scale.is_finite() && offset.is_finite() {
        let normalized = (value / scale) + offset;
        let quantized = normalized.round() as i32;
        return quantized.clamp(min_val, max_val) as i8;
    }
    0i8
}

/// Enhanced dequantization with optional offset support
#[inline]
pub fn dequantize_value_with_offset(quantized: i8, scale: f32, offset: f32) -> f32 {
    if scale.is_finite() && offset.is_finite() {
        (quantized as f32 - offset) * scale
    } else {
        0.0
    }
}

/// Backward-compatible wrapper (maintains current API)
#[inline]
pub fn quantize_value(value: f32, scale: f32, bits: u8) -> i8 {
    quantize_value_with_offset(value, scale, 0.0, bits)
}

/// Backward-compatible wrapper (maintains current API)
#[inline]
pub fn dequantize_value(quantized: i8, scale: f32) -> f32 {
    dequantize_value_with_offset(quantized, scale, 0.0)
}
```

### Calculate Optimal Scale and Offset

```rust
/// Calculate optimal scale and offset for asymmetric quantization
pub fn calculate_scale_offset(data: &[f32], bits: u8) -> (f32, f32) {
    let finite_data: Vec<f32> = data.iter().filter(|&&x| x.is_finite()).copied().collect();
    
    if finite_data.is_empty() {
        return (1.0, 0.0);
    }

    let min_val = finite_data.iter().fold(f32::INFINITY, |a, &b| a.min(b));
    let max_val = finite_data.iter().fold(f32::NEG_INFINITY, |a, &b| a.max(b));
    
    let qmin = -(1 << (bits - 1)) as f32;
    let qmax = ((1 << (bits - 1)) - 1) as f32;
    
    let scale = (max_val - min_val) / (qmax - qmin);
    let offset = qmin - min_val / scale;
    
    // Validate and provide safe fallbacks
    if scale.is_finite() && scale > 0.0 && offset.is_finite() {
        (scale, offset)
    } else {
        (1.0, 0.0)
    }
}
```

### Phase 1: Core Function Enhancement
- [ ] Add `quantize_value_with_offset` and `dequantize_value_with_offset` functions
- [ ] Implement `calculate_scale_offset` utility function
- [ ] Maintain backward compatibility with existing API
- [ ] Add comprehensive input validation and edge case handling

### Phase 2: Integration and Testing
- [ ] Update TL2 quantizer to optionally use offset-based quantization
- [ ] Add comprehensive unit tests for new functions
- [ ] Extend mutation killer tests to cover offset parameter mutations
- [ ] Add property-based tests for asymmetric quantization accuracy

### Phase 3: Performance Optimization
- [ ] Benchmark quantization accuracy improvements with offset support
- [ ] Add SIMD optimizations for bulk quantization operations
- [ ] Implement conditional compilation for advanced quantization features
- [ ] Profile performance impact of offset calculations

### Phase 4: Documentation and Validation
- [ ] Update function documentation with offset parameter usage examples
- [ ] Add cross-validation tests against reference implementations
- [ ] Update quantization accuracy benchmarks
- [ ] Add integration tests with I2_S and TL1 quantizers

### Unit Tests
```rust
#[test]
fn test_asymmetric_quantization_accuracy() {
    // Test data with non-zero mean
    let data = vec![1.0, 2.0, 3.0, 4.0, 5.0];  
    let (scale, offset) = calculate_scale_offset(&data, 2);
    
    for &value in &data {
        let quantized = quantize_value_with_offset(value, scale, offset, 2);
        let dequantized = dequantize_value_with_offset(quantized, scale, offset);
        assert!((dequantized - value).abs() < 0.5); // Better than symmetric
    }
}

#[test]
fn test_backward_compatibility() {
    let test_cases = vec![(1.0, 1.0, 2), (-2.0, 0.5, 2), (0.0, 1.0, 2)];
    
    for (value, scale, bits) in test_cases {
        let old_result = quantize_value(value, scale, bits);
        let new_result = quantize_value_with_offset(value, scale, 0.0, bits);
        assert_eq!(old_result, new_result);
    }
}
```

### Integration Tests
- Cross-validation with GGML reference implementations
- Accuracy comparison: symmetric vs asymmetric quantization
- Performance benchmarks for quantization-sensitive models
- Memory usage validation for scale/offset storage

### Functional Requirements
- [ ] New functions support asymmetric quantization with configurable offset
- [ ] Existing API remains unchanged and fully backward compatible
- [ ] All edge cases handled: NaN, infinity, zero scale, extreme values
- [ ] Quantization accuracy improves for non-zero mean data distributions

### Performance Requirements
- [ ] No performance regression for existing symmetric quantization (offset=0)
- [ ] New functions perform within 5% of current implementation speed
- [ ] Memory overhead for offset storage is minimal (<1% increase)
- [ ] SIMD optimization potential preserved for future enhancements

### Quality Requirements
- [ ] 100% test coverage for new functions and edge cases
- [ ] All existing mutation tests continue to pass
- [ ] New mutation killer tests cover offset parameter mutations
- [ ] Cross-validation tests pass against reference implementations

### Compatibility Requirements
- [ ] Feature flag compatibility maintained (`--features cpu`)
- [ ] No breaking changes to existing quantization interfaces
- [ ] Integration with I2_S, TL1, TL2 quantizers preserved
- [ ] MSRV compatibility (Rust 1.90.0) maintained

## Related Issues

This enhancement supports the broader production readiness initiative:
- **Performance Optimization**: Improved quantization accuracy reduces inference quality degradation
- **Standards Compliance**: Alignment with industry quantization formats (GGML, ONNX)
- **Cross-Validation Framework**: Enhanced testing against reference implementations
- **SIMD Optimization**: Foundation for future vectorized quantization kernels

## Risk Assessment

**Implementation Risks**: Low
- Non-breaking API changes minimize integration disruption
- Comprehensive test coverage prevents regression issues
- Incremental implementation allows for iterative validation

**Performance Risks**: Low
- Offset calculation overhead is minimal for typical use cases
- Backward compatibility ensures no impact on existing workflows
- Optional feature can be disabled if performance issues arise

**Compatibility Risks**: Very Low
- Maintains full backward compatibility with existing API
- Feature-gated implementation allows selective adoption
- Cross-validation framework ensures correctness

## Additional Context

**Code Locations:**
- Primary implementation: `/home/steven/code/Rust/BitNet-rs/crates/bitnet-quantization/src/utils.rs:95-119`
- Usage in TL2: `/home/steven/code/Rust/BitNet-rs/crates/bitnet-quantization/src/tl2.rs:405`
- Test coverage: `/home/steven/code/Rust/BitNet-rs/crates/bitnet-quantization/tests/scale_factor_boundary_mutation_killers.rs:174-189`

**Documentation:**
- Quantization reference: `docs/reference/quantization-support.md`
- Performance guide: `docs/performance-benchmarking.md`
- Architecture overview: `docs/architecture-overview.md`
